[feat] Resubmit the operator module

TimmyExogenous opened the pull request:
## Description

This PR adds an implementation designed in the below file.
[The state update of undelegation and bonding assets](https://www.notion.so/exocore/The-state-update-of-undelegation-and-bonding-assets-653bcfc8a74e4cf9b900976c4ec51078) 

This PR introduces interfaces for recording and updating asset amounts and shares in response to external events, such as delegation/undelegation, opt-in/opt-out, and slashing. The asset shares can also be utilized to calculate the voting power of both stakers and operators. Subsequently, the reward module can calculate rewards based on the computed voting power. 
Additionally, this PR introduces a scheduling function to manage the changes in asset shares caused by price fluctuation. 
All logic has been implemented in a new module named 'operator' in this PR. 

The core logic is located in the following files:
[/x/operator/state_update.go](https://github.com/ExocoreNetwork/exocore/pull/103/files#diff-01d8d01ea401718d80bc903634d5b3cc214bb567b4031337d3502c6368c448a2)
[/x/operator/abci.go](https://github.com/ExocoreNetwork/exocore/pull/103/files#diff-f6b57769ef30f0ab7cc188b1cc1ce7fd7e6a3932cf49ecd5d4aca365178d4ec7)

The provided interfaces are as below:
```go	
UpdateOptedInAssetsState(ctx sdk.Context, stakerId, assetId, operatorAddr string, opAmount sdkmath.Int) error

OptIn(ctx sdk.Context, operatorAddress sdk.AccAddress, AVSAddr string) error

OptOut(ctx sdk.Context, OperatorAddress sdk.AccAddress, AVSAddr string) error

Slash(ctx sdk.Context, operatorAddress sdk.AccAddress, AVSAddr, slashContract, slashId string, occurredSateHeight int64, slashProportion sdkmath.LegacyDec) error
````  
The implementation requires certain interfaces from the Oracle and AVS modules. It needs to retrieve price changes from the Oracle module and access the assets supported by AVS to update the asset shares.

Except for the features mentioned above, this new PR adds the functions  expected by the dogfood module. These functions have been implemented in the app-chain module by Max. The related PR is as follows:
[app-chain PR](https://github.com/ExocoreNetwork/exocore-private/pull/58)

This PR also resolves some problems that would cause lint errors. 
Supersedes [old PR about operator module](https://github.com/ExocoreNetwork/exocore-private/pull/103) from the older repo.

## Todo
* code comments and tests

TimmyExogenous committed:
refactor the non-compliant code

fix the bug caused by rebase merge

remove license comments

add some code comments

change precompile contract solidity version

TimmyExogenous committed:
add more info for error log

TimmyExogenous committed:
add operator module

TimmyExogenous committed:
update avs_operator_assets_state

TimmyExogenous committed:
add opt-out function

TimmyExogenous committed:
update opt-out function

TimmyExogenous committed:
update slash function

TimmyExogenous committed:
update slash function

TimmyExogenous committed:
finish slash function

TimmyExogenous committed:
update abci.go in operator module

TimmyExogenous committed:
finish abci.go in operator module

TimmyExogenous committed:
update abci.go in the delegation module

TimmyExogenous committed:
solve the problems identified by code review

TimmyExogenous committed:
fix some bugs of operator module

TimmyExogenous committed:
resolve the problems of code style

TimmyExogenous committed:
remove the redundant code from the abci.go file in the operator module

TimmyExogenous committed:
change file name in operator module

TimmyExogenous committed:
test getting historical operator info

TimmyExogenous committed:
add ContextForHistoricalState to retrieve historical state

TimmyExogenous committed:
fix the testing bug for retrieving histrical operator info

TimmyExogenous committed:
update for retrieving historical state

TimmyExogenous committed:
fix test bug for delegation

TimmyExogenous committed:
add unit test for operator module

MaxMustermann2 committed:
feat(dogfood): implement sdk staking interface

The interface required by IBC was already implemented in
`validators.go`. This PR takes that a step further and implements the
interfaces required to use the dogfood module as a drop-in replacement
for the SDK's staking module. The interfaces implemented are those
expected by the slashing, evidence and `genutil` modules. An explanation
has been provided above each function to explain why and where it is
called, and if its implementation is really necessary. There are still
some TODOs left in this PR that depend on the overall slashing /
operator design.

This branch is merged into dogfood-part6, which forms the basis of #122.

TimmyExogenous committed:
fix the problems after rebasing to dogfood

TimmyExogenous committed:
copy the code from operator_consent to integrate with the dogfood module.

TimmyExogenous committed:
correct the dependency in x/operator/keeper/utils_test.go

TimmyExogenous committed:
delete snapshot and refine some code about operator module

TimmyExogenous committed:
add unit test for slash in operator module

TimmyExogenous committed:
Test to see if the workflow for checking the PR will be activated.

TimmyExogenous committed:
refine the naming about stakerId assetId and clientChainId

refine the test codes and make all unit tests pass

delete the code related to setting ExoCoreLzAppAddress to fix the bug when set parameter

use shared test utils to refine the test codes

use shared test utils to refine the test codes

fix the problem in tests caused by denom rename

TimmyExogenous committed:
change LayerZeroChainId to LayerZeroChainID

TimmyExogenous committed:
change recordId to recordID

TimmyExogenous committed:
change stakeId to stakeID

TimmyExogenous committed:
Merge branch 'develop' into copy-app-chain

# Conflicts:
#	app/export.go
#	go.mod
#	precompiles/delegation/delegation.sol
#	precompiles/delegation/delegation_test.go
#	precompiles/deposit/deposit.sol
#	precompiles/deposit/types.go
#	precompiles/reward/claimReward.sol
#	precompiles/slash/slash.sol
#	proto/exocore/delegation/v1/query.proto
#	proto/exocore/delegation/v1/tx.proto
#	proto/exocore/restaking_assets_manage/v1/query.proto
#	proto/exocore/restaking_assets_manage/v1/tx.proto
#	testutil/utils.go
#	x/delegation/client/cli/tx.go
#	x/delegation/keeper/abci.go
#	x/delegation/keeper/cross_chain_tx_process.go
#	x/delegation/keeper/delegation_op_test.go
#	x/delegation/keeper/delegation_state.go
#	x/delegation/keeper/grpc_query.go
#	x/delegation/keeper/keeper.go
#	x/delegation/keeper/msg_server.go
#	x/delegation/keeper/operator_info_test.go
#	x/delegation/keeper/un_delegation_state.go
#	x/delegation/types/expected_keepers.go
#	x/delegation/types/keys.go
#	x/delegation/types/query.pb.go
#	x/delegation/types/query.pb.gw.go
#	x/delegation/types/tx.pb.go
#	x/deposit/types/deposit.pb.go
#	x/deposit/types/query.pb.go
#	x/deposit/types/tx.pb.go
#	x/dogfood/keeper/impl_sdk.go
#	x/dogfood/types/expected_keepers.go
#	x/native_token/types/tx.pb.go
#	x/operator/types/codec.go
#	x/restaking_assets_manage/client/cli/query.go
#	x/restaking_assets_manage/genesis.go
#	x/restaking_assets_manage/keeper/client_chain_asset.go
#	x/restaking_assets_manage/keeper/exocore_addr.go
#	x/restaking_assets_manage/keeper/grpc_query.go
#	x/restaking_assets_manage/keeper/operator_asset.go
#	x/restaking_assets_manage/keeper/staker_asset.go
#	x/restaking_assets_manage/types/genesis.pb.go
#	x/restaking_assets_manage/types/keys.go
#	x/restaking_assets_manage/types/query.pb.go
#	x/restaking_assets_manage/types/query.pb.gw.go
#	x/restaking_assets_manage/types/tx.pb.go
#	x/reward/keeper/claim_reward.go
#	x/reward/types/genesis.pb.go
#	x/reward/types/params.pb.go
#	x/reward/types/query.pb.go
#	x/reward/types/tx.pb.go
#	x/reward/types/types.pb.go
#	x/slash/keeper/execute_slash.go
#	x/slash/types/genesis.pb.go
#	x/slash/types/params.pb.go
#	x/slash/types/query.pb.go
#	x/slash/types/tx.pb.go
#	x/withdraw/client/cli/tx.go
#	x/withdraw/keeper/claim_withdraw.go
#	x/withdraw/module.go

TimmyExogenous committed:
fix super lint error
update the dependency version of protobuf to eliminate vulnerability
fix the lint error about gosec and vulncheck
fix the lint error about proto-lint

TimmyExogenous commented:
The failing "Protobuf / break-check" workflow can be ignored. It reminds us that some protobuf files have been changed. But It's okay to change them during the early stages of rapid iteration and development. This check won't fail after It's been merged.

rlan35 commented on proto/exocore/delegation/v1/query.proto:
```diff
@@ -20,8 +19,8 @@ message DelegationInfoReq {
-  // can_undelegation_amount is the amount that can be undelegated.
-  string can_undelegation_amount = 1
+  // can_be_undelegated_amount is the amount that can be undelegated.
+  string can_be_undelegated_amount = 1
```
undelegatable_amount

TimmyExogenous replied:
okay

rlan35 commented on proto/exocore/delegation/v1/query.proto:
```diff
@@ -34,6 +33,13 @@ message DelegationAmounts {
     (gogoproto.nullable) = false
   ];
+  // can_be_undelegated_after_slash is the amount that can be undelegated after slash
+  string can_be_undelegated_after_slash = 3
```
undelegatable_amount_after_slash

rlan35 commented on proto/exocore/operator/v1/tx.proto:
```diff
@@ -0,0 +1,166 @@
+  ];
+}
+
+// ClientChainEarningAddrList is the list of client chain earning addresses.
```
what is a earning address? Please add comments here.

rlan35 commented on proto/exocore/operator/v1/tx.proto:
```diff
@@ -0,0 +1,166 @@
+  // slash_height is the height at which the slash event is submitted
+  int64 slash_height = 2;
+  // occurred_height is the height at which the slash event occurs
+  int64 occurred_height = 3;
```
=> event_height

rlan35 commented on proto/exocore/operator/v1/tx.proto:
```diff
@@ -0,0 +1,166 @@
+  // occurred_height is the height at which the slash event occurs
+  int64 occurred_height = 3;
+  // execute_height is the height at which the slash event is executed
+  int64 execute_height = 4;
```
why the slash_height and execute_height is different? In that case, better name like this
submitted_height
processed_height
event_height

MaxMustermann2 replied:
Yes, please clarify the height names via comments as well. I am guessing that these are heights of the Exocore chain but it helps to write it down too.

TimmyExogenous replied:
Yeah, I will rename them and add some comments.

rlan35 commented on proto/exocore/operator/v1/tx.proto:
```diff
@@ -0,0 +1,166 @@
+  // execute_height is the height at which the slash event is executed
+  int64 execute_height = 4;
+  // is_veto is a flag to indicate if this slash is vetoed
+  bool is_veto = 5;
```
is_vetoed

rlan35 commented on proto/exocore/restaking_assets_manage/v1/tx.proto:
```diff
@@ -29,6 +40,20 @@ message ClientChainInfo {
+  // chain_id is used as the primary key
+  string chain_id = 3;
+  // exo_core_chain_index is the index of the chain in exocore, so far unused
+  uint64 exo_core_chain_index = 4;
```
Exocore is a single word. It's not two words. Should be exocore_chain_index

MaxMustermann2 replied:
It was originally written as ExoCoreChainIndex and I changed it accordingly in one of the linting PRs. Please correct it to one word.

TimmyExogenous replied:
okay

rlan35 commented on proto/exocore/restaking_assets_manage/v1/tx.proto:
```diff
@@ -88,9 +113,9 @@ message StakerSingleAssetOrChangeInfo {
+  // wait_unbonding_amount_or_want_change_value is the amount that is waiting for undelegation
   // or the amount by which it can change.
-  string wait_undelegation_amount_or_want_change_value = 3
+  string wait_unbonding_amount_or_want_change_value = 3
```
Still mixing two purposes in the same field? Better separate them.

TimmyExogenous replied:
Yeah, I will refine them in the next commit.

rlan35 commented on proto/exocore/restaking_assets_manage/v1/tx.proto:
```diff
@@ -125,9 +151,27 @@ message OperatorSingleAssetOrChangeInfo {
+
+  // operator_own_wait_unbonding_amount is the amount that is owned by operator itself and waiting for unbonding
+  // or the amount by which it can change
+  string operator_own_wait_unbonding_amount = 4
```
operator_unbonding_amount

rlan35 commented on proto/exocore/restaking_assets_manage/v1/tx.proto:
```diff
@@ -125,9 +151,27 @@ message OperatorSingleAssetOrChangeInfo {
+
+  // operator_can_unbond_after_slash is the amount that is owned by operator itself and can be unbonded after slash
+  // or the amount by which it can change
+  string operator_can_unbond_after_slash = 5
```
operator_unbondable_amount_after_slash

MaxMustermann2 commented on precompiles/deposit/deposit.sol:
```diff
@@ -1,6 +1,6 @@
 pragma solidity >=0.8.17;
 
-/// @dev The DEPOSIT contract's address.
+/// @dev The DEPOSIT contract's.Address.
```
Nit: please undo this change.

TimmyExogenous replied:
Thanks

MaxMustermann2 commented on precompiles/deposit/deposit.sol:
```diff
@@ -15,8 +15,8 @@ IDeposit constant DEPOSIT_CONTRACT = IDeposit(
 /// @dev deposit the client chain assets to the staker, that will change the state in deposit module
-/// Note that this address cannot be a module account.
-/// @param clientChainLzID The lzId of client chain
+/// Note that this.Address cannot be a module account.
```
Nit: please undo this change.

MaxMustermann2 commented on proto/exocore/operator/v1/query.proto:
```diff
@@ -0,0 +1,47 @@
+  string chain_id = 2;
+}
+
+// QueryOperatorConsKeyResponse is the response for QueryOperatorConsKeyForChainID
```
`QueryOperatorConsKeyForChainID` -> `QueryOperatorConsKeyRequest`

MaxMustermann2 commented on proto/exocore/operator/v1/tx.proto:
```diff
@@ -0,0 +1,166 @@
+
+// DecValueField is a field that holds a value of sdk.LegacyDec type.
+message DecValueField {
+  // amount is the USD value of the asset, as an sdk.LegacyDec.
```
Is it always the USD value or is it the quantity of the asset that needs to be multiplied with the exchange rate?

TimmyExogenous replied:
It's the asset's USD value that has been multiplied by the exchange rate.

MaxMustermann2 commented on proto/exocore/operator/v1/tx.proto:
```diff
@@ -0,0 +1,166 @@
+}
+
+// OptedInAssetState is the state of opted-in asset
+message OptedInAssetState{
```
Nit: Please add a space before the `{`.

MaxMustermann2 commented on proto/exocore/operator/v1/tx.proto:
```diff
@@ -0,0 +1,166 @@
+message RegisterOperatorResponse{}
+
+// OptInToChainIDRequest defines the OptInToChainID request.
+message OptInToChainIDRequest {
```
Consider renaming these from `ChainID` to `CosmosChain`. An operator can opt into a cosmos chain via these, which may be an app chain or the Exocore chain itself through the dogfood module.

TimmyExogenous replied:
Yeah, using OptInToCosmosChainRequest is better.

MaxMustermann2 commented on proto/exocore/operator/v1/tx.proto:
```diff
@@ -0,0 +1,166 @@
+
+
+
+
```
Nit: Please remove the blank lines.

TimmyExogenous replied:
Have been resolved

MaxMustermann2 commented on testutil/abci.go:
```diff
@@ -32,10 +32,10 @@ func Commit(ctx sdk.Context, app *app.ExocoreApp, t time.Duration, vs *tmtypes.V
 }
 
-// CommitAndCreateNewCtx commits a block at a given time creating a ctx with the current settings
+// CommitAndCreateNewCtx commits a block at a given time creating a Ctx with the current settings
```
nit: revert

MaxMustermann2 commented on testutil/abci.go:
```diff
@@ -44,8 +44,14 @@ func CommitAndCreateNewCtx(ctx sdk.Context, app *app.ExocoreApp, t time.Duration
-	newCtx := app.BaseApp.NewContext(false, header)
-	// set the reseted fields to keep the current ctx settings
+	var newCtx sdk.Context
+	if isUncachedCtx {
```
`isUncachedCtx` -> `useUncachedCtx`

MaxMustermann2 commented on testutil/abci.go:
```diff
@@ -222,7 +228,7 @@ func commit(ctx sdk.Context, app *app.ExocoreApp, t time.Duration, vs *tmtypes.V
 }
 
-// checkTxBytes encodes a transaction and calls checkTx on the app.
+// checkTxBytes encodes a transaction and calls checkTx on the App.
```
nit: undo.

MaxMustermann2 commented on x/delegation/keeper/abci.go:
```diff
@@ -24,58 +20,80 @@ func (k Keeper) EndBlock(ctx sdk.Context, _ abci.RequestEndBlock) []abci.Validat
+		// any special handling here. if an operator undelegates before they opt out, the undelegation
+		// will be processed normally. if they undelegate after they opt out, the undelegation will
+		// be released at the same time as opt out completion, provided there are no other chains that
+		// the operator is still active on. the same applies to delegators too.
```
Is this comment still true based on the opt out design changes made since then?

TimmyExogenous replied:
This comment can be removed. The undelegation won't be influenced by the opt-out.

MaxMustermann2 commented on x/delegation/keeper/cross_chain_tx_process.go:
```diff
@@ -109,10 +109,10 @@ type DelegationOrUndelegationParams struct {
 
 // DelegateTo : It doesn't need to check the active status of the operator in middlewares when delegating assets to the operator. This is because it adds assets to the operator's amount. But it needs to check if operator has been slashed or frozen.
-func (k Keeper) DelegateTo(ctx sdk.Context, params *DelegationOrUndelegationParams) error {
+func (k *Keeper) DelegateTo(ctx sdk.Context, params *DelegationOrUndelegationParams) error {
```
Why is the function receiver a pointer now?

TimmyExogenous replied:
It's for the consistency. If some of the methods of the type must have pointer receivers, the rest should too. This is a code convention described in the file linked below.
[methods_on_values_or_pointers](https://go.dev/doc/faq#methods_on_values_or_pointers) 
The Goland IDE reminds me of it. Although the cosmos-SDK doesn't adhere to this convention.

MaxMustermann2 replied:
I see; it makes sense. Can you tell me which methods must have pointer receivers and why?

MaxMustermann2 replied:
The pointer receiver is required to `SetHooks` on the keeper. At other places, it is due to convention.

MaxMustermann2 commented on x/delegation/keeper/cross_chain_tx_process.go:
```diff
@@ -164,15 +166,23 @@ func (k Keeper) DelegateTo(ctx sdk.Context, params *DelegationOrUndelegationPara
 		return err
 	}
+	// call operator module to bond the increased assets to the opted-in AVS
+	err = k.expectedOperatorInterface.UpdateOptedInAssetsState(ctx, stakerID, assetID, params.OperatorAddress.String(), params.OpAmount)
```
I think this needs to be abstracted away from this module to keep them independent. The delegation module should ideally not care or know about the data storage mechanism followed by the operator module. Instead of calling the `UpdateOptedInAssetsState`, it would be better to have the operator keeper implement the delegation hooks (with the addition of new parameters which it does not yet accept). Within those hooks, it can take the necessary action. What do you think?

TimmyExogenous replied:
Yeah, implementing through hooks is definitely a better way.  This might be refined in the future PR.

MaxMustermann2 commented on x/delegation/keeper/cross_chain_tx_process.go:
```diff
@@ -199,38 +210,55 @@ func (k Keeper) UndelegateFrom(ctx sdk.Context, params *DelegationOrUndelegation
 	}
+
+	// call operator module to decrease the state of opted-in assets
+	err = k.expectedOperatorInterface.UpdateOptedInAssetsState(ctx, stakerID, assetID, params.OperatorAddress.String(), params.OpAmount.Neg())
```
Same comment as above to move it within the hooks and abstract it away from the delegation module.

TimmyExogenous replied:
Same answer as above.

MaxMustermann2 commented on x/delegation/keeper/delegation_state.go:
```diff
@@ -40,68 +38,76 @@ func (k Keeper) UpdateStakerDelegationTotalAmount(ctx sdk.Context, stakerID stri
-	if !store.Has(prefixKey) {
+	prefixKey := stakingtypes.GetJoinedStoreKey(stakerID, assetID)
+	isExit := store.Has(prefixKey)
+	if !isExit {
```
Remove `isExit` since it is not used again.
```suggestion
	if !store.Has(prefixKey)
```

TimmyExogenous replied:
Thanks

MaxMustermann2 commented on x/delegation/keeper/delegation_state.go:
```diff
@@ -141,3 +145,58 @@ func (k Keeper) GetDelegationInfo(ctx sdk.Context, stakerID, assetID string) (*d
 	return &ret, nil
 }
+
+// DelegationStateByOperatorAssets get the specified assets state delegated to the specified operator
```
Please describe in the comment here the return mapping.

MaxMustermann2 commented on x/delegation/keeper/delegation_state.go:
```diff
@@ -141,3 +145,58 @@ func (k Keeper) GetDelegationInfo(ctx sdk.Context, stakerID, assetID string) (*d
+			return err
+		}
+		if len(keys) == 3 {
+			err = f(keys[0], keys[1], keys[2], &amounts)
```
Would it be better to make this function return a boolean value instead of error? I have seen boolean values being used in the Cosmos SDK and not error types in most situations. If the boolean returned is false, the iteration exits. Here's an example from the `x/staking` module.

```go
func (k Keeper) IterateDelegations(ctx sdk.Context, delAddr sdk.AccAddress,
	fn func(index int64, del types.DelegationI) (stop bool),
) {
	store := ctx.KVStore(k.storeKey)
	delegatorPrefixKey := types.GetDelegationsKey(delAddr)

	iterator := sdk.KVStorePrefixIterator(store, delegatorPrefixKey) // smallest to largest
	defer iterator.Close()

	for i := int64(0); iterator.Valid(); iterator.Next() {
		del := types.MustUnmarshalDelegation(k.cdc, iterator.Value())

		stop := fn(i, del)
		if stop {
			break
		}
		i++
	}
}
```

TimmyExogenous replied:
I see the a lot of interfaces of state implemented in cosmos-sdk  don't have an error as a return value. I think if `IterateDelegations` returns an error,   the `fn` might also need to return an error to handle the situation in which  the `fn` may cause an error.  But I am not sure whether we should change our interfaces according to the cosmos-sdk.
So maybe I can change it when we figure out whether the implementation of cosmos-sdk is better.

MaxMustermann2 commented on x/delegation/keeper/keeper.go:
```diff
@@ -21,10 +18,12 @@ type Keeper struct {
+	restakingStateKeeper      keeper.Keeper
+	depositKeeper             depositkeeper.Keeper
+	slashKeeper               delegationtype.ISlashKeeper
+	expectedOperatorInterface delegationtype.ExpectedOperatorInterface
```
Please rename this struct field to `operatorKeeper`, in line with the other members.

TimmyExogenous replied:
Thanks

MaxMustermann2 commented on x/delegation/keeper/keeper.go:
```diff
@@ -21,10 +18,12 @@ type Keeper struct {
-	operatorOptedInKeeper delegationtype.OperatorOptedInMiddlewareKeeper
+	// other keepers
+	restakingStateKeeper      keeper.Keeper
+	depositKeeper             depositkeeper.Keeper
```
Convert to the minimum required interface instead of directly linking to the module's keeper. This mechanism is used to avoid import cycles in the future.

TimmyExogenous replied:
Yeah, thanks

MaxMustermann2 commented on x/delegation/keeper/un_delegation_state.go:
```diff
@@ -43,15 +44,23 @@ func (k Keeper) SetUndelegationRecords(ctx sdk.Context, records []*types.Undeleg
 }
 
-func (k Keeper) GetUndelegationRecords(ctx sdk.Context, singleRecordKeys []string, getType GetUndelegationRecordType) (record []*types.UndelegationRecord, err error) {
+// StoreWaitCompleteRecord add it to handle the delay of completing undelegation caused by onHoldCount
```
Please update this comment for clarity. `StoreWaitCompleteRecord is used to update only the undelegation record and no ancillary information. In the event that the undelegation is held by another module, this function is used within the EndBlocker to increment the scheduled completion block number by 1.`

TimmyExogenous replied:
Yeah

MaxMustermann2 commented on x/delegation/keeper/un_delegation_state.go:
```diff
@@ -141,3 +150,29 @@ func (k Keeper) GetWaitCompleteUndelegationRecords(ctx sdk.Context, height uint6
+func (k *Keeper) IncrementUndelegationHoldCount(ctx sdk.Context, recordKey []byte) {
+	prev := k.GetUndelegationHoldCount(ctx, recordKey)
+	if prev == math.MaxUint64 {
+		panic("cannot increment undelegation hold count above max uint64")
```
It is better to remove the `panic`s happening here and instead return an error.

TimmyExogenous replied:
Yeah, the hook interface might also need to return an error.

MaxMustermann2 commented on x/delegation/types/expected_keepers.go:
```diff
@@ -15,20 +15,31 @@ type ISlashKeeper interface {
 
-type OperatorOptedInMiddlewareKeeper interface {
-	GetOperatorCanUndelegateHeight(ctx sdk.Context, assetID string, opAddr sdk.AccAddress, startHeight uint64) uint64
+// DelegationHooks add for dogfood
```
`DelegationHooks add for dogfood` => please change this because the delegation hooks are planned to be used in other modules as well. You can explain it as `DelegationHooks are event hooks triggered by the delegation module`.

TimmyExogenous replied:
thanks

MaxMustermann2 commented on x/delegation/types/expected_keepers.go:
```diff
@@ -15,20 +15,31 @@ type ISlashKeeper interface {
 }
 
-type VirtualOperatorOptedInKeeper struct{}
+type ExpectedOperatorInterface interface {
```
Please rename to `OperatorKeeper`

MaxMustermann2 commented on x/delegation/types/expected_keepers.go:
```diff
@@ -15,20 +15,31 @@ type ISlashKeeper interface {
 // VirtualISlashKeeper todo: When the actual keeper functionality has not been implemented yet, temporarily use the virtual keeper.
```
Please add interface guards for any virtual keeper implementations being used here.

MaxMustermann2 commented on x/operator/keeper/abci.go:
```diff
@@ -0,0 +1,135 @@
+package keeper
```
Are we updating the shares at each block? The dogfood module is designed to handle this only on a per-epoch basis to lower the amount of processing required every 3 seconds.

TimmyExogenous replied:
Yeah, the shares will be updated each block. I saw the dogfood module is designed to update states only on a per-epoch basis. In terms of the functionality, I don't think it will have any impact on dogfood. In the future, if performance issues are considered, epoch may be introduced into the operator module.

MaxMustermann2 replied:
Has this approach been discussed with RJ? Updating each block is very cost heavy.

TimmyExogenous replied:
Yeah, this has been discussed with RJ. The first version is just a version to implement the basic features. The epoch will be addressed in the next version. 
In the current implementation, the assets' amount and share of stakers and operators will be updated immediately when the related delegation/undelegation, slash, opt-in/out and price change occur.  The share update caused by the price change will be handled on a per-epoch basis after the code is implemented with Oracle hook.  As for the other situation, we might keep the update of assets' amount unchanged but change the share update to be on a per-epoch basis.

MaxMustermann2 commented on x/operator/keeper/avs_operator_shares.go:
```diff
@@ -0,0 +1,282 @@
+package keeper
```
Please add comments for each function.

MaxMustermann2 replied:
Within this file, the `avsAddr` is a string. What is the format (sorry if I missed any document that explained it) ? Can the same string be used for app chains, or should we build separate functions?

TimmyExogenous replied:
The `avsAddr` is an ID to identify an Avs. The format hasn't been determined. Maybe we can use `prefix+type+exoAddr`, if the type is an app chain then the chainID might be added as a suffix. I think it should be defined in the Avs module that is responsible for managing all the Avs, such as registering or updating information.

MaxMustermann2 commented on x/operator/keeper/dogfood.go:
```diff
@@ -0,0 +1,425 @@
+package keeper
```
Instead of calling this file `dogfood.go`, please rename it to `consensus_keys.go` or similar. These keys will be used by app chains as well as dogfood.

TimmyExogenous replied:
Yeah

MaxMustermann2 commented on x/operator/keeper/keeper.go:
```diff
@@ -0,0 +1,75 @@
+	cdc      codec.BinaryCodec
+
+	// other keepers
+	restakingStateKeeper keeper.Keeper
```
Turn this to an interface as well.

MaxMustermann2 commented on x/operator/keeper/msg_server.go:
```diff
@@ -0,0 +1,75 @@
+	return nil, nil
+}
+
+// OptInToChainID add for dogfood
```
Update the comment.

MaxMustermann2 commented on x/operator/keeper/operator.go:
```diff
@@ -0,0 +1,117 @@
+
+// SetOperatorInfo This function is used to register to be an operator in exoCore, the provided info will be stored on the chain.
+// Once an address has become an operator,the operator can't return to a normal address.But the operator can update the info through this function
+// As for the operator opt-in function,it needs to be implemented in operator opt-in or AVS module
```
This last line can be deleted right?

TimmyExogenous replied:
thanks

MaxMustermann2 commented on x/operator/keeper/operator.go:
```diff
@@ -0,0 +1,117 @@
+	return true
+}
+
+func (k *Keeper) GetOptedInAVSForOperator(ctx sdk.Context, operatorAddr string) ([]string, error) {
```
This function should also return the app chains / dogfood chain?

TimmyExogenous replied:
Yeah, It returns a slice of `avsAddr`. As mentioned above, the `avsAddr` should be a universal identifier for all AVS, including app chain AVS.

MaxMustermann2 commented on x/operator/keeper/state_update.go:
```diff
@@ -0,0 +1,494 @@
+			}
+		}
+	}
+	return nil
```
A lot of information is being stored by this function. Where is it used? Is all of it required?

TimmyExogenous replied:
In the operator module, the USD share states related to assets, stakers, operators, and AVS will be stored separately. This function will update the USD share states related to assets, stakers, operators, and AVS when the asset amount changes due to delegation, undelegation, slashStaker, and slashOperator. In the current implementation, the stored states haven't been uniformly optimized, so all of them are required

MaxMustermann2 replied:
I think we should work on this design.

TimmyExogenous replied:
Yeah, the data structure of the stored information in the assets, delegation and operator modules should be redesigned and uniformly optimized. Maybe we can work on it after all the basic features have been finished.

MaxMustermann2 commented on x/operator/keeper/state_update.go:
```diff
@@ -0,0 +1,494 @@
+}
+
+// GetAssetsAmountToSlash It will slash the assets that are opting into AVS first, and if there isn't enough to slash, then it will slash the assets that have requested to undelegate but still locked.
+func (k *Keeper) GetAssetsAmountToSlash(ctx sdk.Context, operatorAddress sdk.AccAddress, avsAddr string, occurredSateHeight int64, slashProportion sdkmath.LegacyDec) (*SlashAssets, error) {
```
What is `occurredSateHeight` ? Looks like a typo.

TimmyExogenous replied:
It's the height at which the slash event occurs, I will rename it to `slashEventHeight`.

MaxMustermann2 commented on x/operator/types/expected_keepers.go:
```diff
@@ -0,0 +1,99 @@
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type ExpectDelegationInterface interface {
```
Please rename to `DelegationKeeper`

TimmyExogenous replied:
Thanks

MaxMustermann2 commented on x/operator/types/expected_keepers.go:
```diff
@@ -0,0 +1,99 @@
+	NewPrice      sdkmath.Int
+	Decimal       uint8
+}
+type ExpectOracleInterface interface {
```
Please rename to `OracleKeeper` and add comments describing the expected interface because I don't see these functions in the oracle PR or design.

TimmyExogenous replied:
Thanks

MaxMustermann2 commented on x/operator/types/expected_keepers.go:
```diff
@@ -0,0 +1,99 @@
+	GetPriceChangeAssets(ctx sdk.Context) (map[string]*PriceChange, error)
+}
+
+type MockOracle struct{}
```
Interface guard.

MaxMustermann2 commented on x/operator/types/expected_keepers.go:
```diff
@@ -0,0 +1,99 @@
+	usdtAssetID := "0xdac17f958d2ee523a2206206994597c13d831ec7_0x65"
+	ret[usdtAssetID] = &PriceChange{
+		NewPrice:      sdkmath.NewInt(1),
+		OriginalPrice: sdkmath.NewInt(1),
```
Is the expectation that the original price as of the last block or as of genesis?

TimmyExogenous replied:
The expected price change will be fully controlled by the Oracle module.  The scheduled time for providing the price change might be configured by the parameter of Oracle. The change should be calculated compared to the previous price stored in the Oracle module. This hasn't been determined yet, the price changes state of assets can also be stored and updated in the operator module.

MaxMustermann2 replied:
Based on the current usage of this function, it appears to expect the price change to be as of each block. The oracle module's implementation may cause some of this to change.

TimmyExogenous replied:
Have discussed this with Leon. The Oracle module will define a hook for this, then the operator module needs to implement the hook. The current implementation here definitely needs to be changed after the hook is defined by the  Oracle module.

MaxMustermann2 commented on proto/exocore/operator/v1/query.proto:
```diff
@@ -0,0 +1,47 @@
+
+// QueryOperatorInfoReq is the request to obtain the operator information.
+message GetOperatorInfoReq {
+  // operator_addr is the operator address.
```
To confirm, this is the `sdk.AccAddress` right? If yes, please add a comment to prevent confusion with traditional Cosmos SDK where the operator's address is generally of the type `sdk.ValAddress`.

TimmyExogenous replied:
Yeah,  I will add a comment

MaxMustermann2 commented on x/delegation/keeper/cross_chain_tx_process.go:
```diff
@@ -109,10 +109,10 @@ type DelegationOrUndelegationParams struct {
-	if !k.IsOperator(ctx, params.OperatorAddress) {
-		return delegationtype.ErrOperatorNotExist
+	if !k.expectedOperatorInterface.IsOperator(ctx, params.OperatorAddress) {
+		return errorsmod.Wrap(delegationtype.ErrOperatorNotExist, fmt.Sprintf("input opreatorAddr is:%s", params.OperatorAddress))
```
Typo: opreator

TimmyExogenous replied:
Thanks

MaxMustermann2 commented on x/delegation/keeper/delegation_op_test.go:
```diff
@@ -38,15 +43,15 @@ func (suite *DelegationTestSuite) TestDelegateTo() {
 	}
 	err = suite.App.DelegationKeeper.DelegateTo(suite.Ctx, delegationParams)
-	suite.EqualError(err, delegationtype.ErrOperatorNotExist.Error())
+	suite.EqualError(err, errorsmod.Wrap(delegationtype.ErrOperatorNotExist, fmt.Sprintf("input opreatorAddr is:%s", delegationParams.OperatorAddress)).Error())
```
typo: opreator

TimmyExogenous replied:
thanks

MaxMustermann2 commented on x/restaking_assets_manage/types/keys.go:
```diff
@@ -81,40 +88,64 @@ var (
+}
+
+// add for dogfood
+func OperatorSnapshotKey(operatorAddr sdk.AccAddress, height uint64) []byte {
```
These snapshot functions can be removed.

TimmyExogenous replied:
Okay

MaxMustermann2 commented on x/operator/client/cli/query.go:
```diff
@@ -0,0 +1,56 @@
+func GetQueryCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:                        operatortypes.ModuleName,
+		Short:                      "Querying commands for the delegation module",
```
`for the operator module`

MaxMustermann2 commented on x/operator/client/cli/tx.go:
```diff
@@ -0,0 +1,81 @@
+func NewTxCmd() *cobra.Command {
+	txCmd := &cobra.Command{
+		Use:                        operatortypes.ModuleName,
+		Short:                      "delegation subcommands",
```
operator subcommands

TimmyExogenous replied:
Thanks

MaxMustermann2 commented on x/operator/keeper/operator_slash_state.go:
```diff
@@ -0,0 +1,115 @@
+package keeper
```
Please add comments for these functions, which should include when they are called by which module.

TimmyExogenous replied:
Okay

MaxMustermann2 commented on x/operator/types/expected_keepers.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+type ExpectAvsInterface interface {
+	GetAvsSupportedAssets(ctx sdk.Context, avsAddr string) (map[string]interface{}, error)
```
In the private repo you said that the `ctx` must be historical (as of the slash infraction height). Please add a comment to that effect here.

TimmyExogenous replied:
The `ctx` can be historical or current, depending on the state you wish to retrieve.  I will add a comment here.

MaxMustermann2 commented on x/operator/keeper/state_update.go:
```diff
@@ -0,0 +1,494 @@
+// OptOut call this function to opt out of AVS
+func (k *Keeper) OptOut(ctx sdk.Context, operatorAddress sdk.AccAddress, avsAddr string) error {
+	// check optedIn info
+	if !k.IsOptedIn(ctx, operatorAddress.String(), avsAddr) {
```
In the private repo, we discussed as follows:

> This function's contents should not be triggered immediately on opt out. They should be done after the unbonding period on the AVS expires.

> Do you mean that the opt-out of AVS also requires a lock period? The undelegation of a staker definitely needs an unbonding period for slashing challenges. In the implementation, assets still need to wait a period to unbond when an undelegation is requested, even if the related operator has already opted out from the AVS. Therefore, the opt-out might not need an unbonding period.

You are correct. However, is there any data that you might wish to retain until the opt out start time + unbonding period finish ? For example, in the case of app chains / dogfood I am retaining the mapping of consensus public key to operator address. When slashing is triggered, the consensus key is supplied (by the SDK) as a parameter, from which I look up the operator address to slash. If I were to delete this data immediately upon opt out, such a look up would not be possible.

TimmyExogenous replied:
It's to be determined whether the opt-out of AVS also requires a lock period.  In the current implementation of the operator module, the lock period of opt-out hasn't been considered. But I also noticed that the opt-out design in the dogfood has a lock period, this is an inconstancy that needs to be considered. If we don't set a lock period for the opt-out, about the state that will be deleted immediately when the AVS opts out, can we preserve it until the slash challenge period expires? That is to say, we can let the system delete the historical state according to the `PruningOptions`

MaxMustermann2 replied:
You are right that the dogfood module can use the historical context to determine the operator address using the consensus public key as an input, which can then be used for slashing.

However, in the undelegation hook, the code also uses the opt-out status of the operator (whether it is opting in or in the process of opting out) to determine the `unbondingCompletionEpoch`. This epoch is when the `DecrementUndelegationHoldCount` is called.

In other words, the undelegation matures either at the operator's opt out completion epoch, or the _usual_ undelegation completion epoch, whichever is earlier. I can store the former within the dogfood module and the latter can be calculated so I don't need an opt-out lock-in period implemented.

MaxMustermann2 replied:
@TimmyExogenous I have another question about this. If the operator opts in, waits for a while, opts out and then immediately opts back in, is this supported?

TimmyExogenous replied:
The implemented `OptIn` and `OptOut` functionalities seem like they can support this. I didn't consider this situation when writing these codes. However, in the implementation, the related state will be updated immediately when the `OptIn` and `OptOut` are called. What potential issues might arise that are not supported?

MaxMustermann2 replied:
There are no immediate concerns that I can think of. Previously, I had used a queue implementation in the dogfood module which had made certain assumptions about the operator module (but this feature was still supported).

TimmyExogenous committed:
refine the code related to updating state

rename package restaking_assets_manage to assets

refine some code related to naming and comments

mikebraver commented on x/delegation/keeper/delegation_op_test.go:
```diff
@@ -38,15 +43,15 @@ func (suite *DelegationTestSuite) TestDelegateTo() {
+	suite.EqualError(err, errorsmod.Wrap(delegationtype.ErrOperatorNotExist, fmt.Sprintf("input opreatorAddr is:%s", delegationParams.OperatorAddress)).Error())
 
-	registerReq := &delegationtype.RegisterOperatorReq{
+	registerReq := &types2.RegisterOperatorReq{
```
rename `type2` to better name?

TimmyExogenous replied:
Yeah, I will change it.

MaxMustermann2 commented on proto/exocore/operator/v1/tx.proto:
```diff
@@ -77,14 +79,14 @@ message OptedInAssetState{
+  // event_height is the exocore block height at which the slash event occurs
+  int64 event_height = 3;
+  // processed_height is the exocore block height at which the slash event is processed
+  int64 processed_height = 4;
```
What would cause `processed_height` to be different from `submitted_height` ?

TimmyExogenous replied:
When considering the veto function, there might be a waiting period from the `submitted_height`. The state of slash assets will be sent to the client chain at `processed_height` once the veto period expires.

MaxMustermann2 commented on x/assets/types/general.go:
```diff
@@ -43,6 +35,30 @@ const (
+
+// StakerSingleAssetChangeInfo This is a struct to describe the desired change that matches with the StakerSingleAssetInfo
+type StakerSingleAssetChangeInfo struct {
+	ChangeForTotalDeposit  math.Int
```
Given that the structure is already named `ChangeInfo`, the `ChangeFor` prefix on its members is redundant.

TimmyExogenous replied:
Yeah, I will rename them.

MaxMustermann2 reviewed (changes requested):
I like the last commit.

TimmyExogenous committed:
redefine change info struct and fix some lint problems

TimmyExogenous committed:
Merge branch 'develop' into copy-app-chain

# Conflicts:
#	go.mod
#	go.sum

TimmyExogenous committed:
mv the function related to setting ExocoreLzAppAddress to assets module and delete redudant code

rename chainId to chainID in dogfood module

rename restakingKeeper to assetsKeeper

TimmyExogenous commented:
All problems identified by code review have been resolved.

rlan35 commented on proto/exocore/assets/v1/query.proto:
```diff
@@ -49,7 +49,7 @@ message QueryStakerAssetInfo {
 message QueryAssetInfoResponse {
   // asset_infos is the response for the staker asset info, indexed by the asset id.
-  map<string, StakerSingleAssetOrChangeInfo> asset_infos = 1;
+  map<string, StakerSingleAssetInfo> asset_infos = 1;
```
StakerAssetInfo     "single" is implied and is redundant.

rlan35 commented on proto/exocore/assets/v1/tx.proto:
```diff
@@ -102,32 +123,47 @@ message StakerSingleAssetOrChangeInfo {
-  // operator_own_amount_or_want_change_value is the amount that the operator owns
-  // or the amount by which it can change.
+
+  // operator_own_amount is the amount that the operator owns.
```
Same here. operator_amount is cleaner. it's implied that it's operator's (own) amount since you already qualified the "amount" with "operator"

rlan35 commented on x/assets/keeper/operator_asset.go:
```diff
@@ -0,0 +1,117 @@
+		}
+		if len(keys) == 3 {
+			err = f(keys[0], keys[1], &amounts)
+			if err != nil {
```
since err is created from f(), is it possible to be nil? If not, just do 

return f(keys[0], keys[1], &amounts)

TimmyExogenous replied:
The return error is possible to be nil.

TimmyExogenous replied:
This shouldn't be changed.

rlan35 commented on x/assets/types/errors.go:
```diff
@@ -20,4 +20,8 @@ var (
 
 	ErrInputPointerIsNil = errorsmod.Register(ModuleName, 7, "the input pointer is nil")
+
+	ErrOperatorAddr = errorsmod.Register(ModuleName, 8, "the operator address isn't a valid acc addr")
```
ErrInvalidOperatorAddr  (naming should be self explanatory)   "the operator address isn't a valid account address" (message can be verbose, no need to use abbreviation)

rlan35 commented on x/assets/types/errors.go:
```diff
@@ -20,4 +20,8 @@ var (
+
+	ErrOperatorAddr = errorsmod.Register(ModuleName, 8, "the operator address isn't a valid acc addr")
+
+	ErrNoAppChainKey = errorsmod.Register(ModuleName, 10, "there is no stored key for the input app chain id")
```
ErrUnknownAppChainID    "the app chain id is unknown or invalid"

rlan35 commented on x/assets/types/keys.go:
```diff
@@ -81,40 +88,64 @@ var (
-	// KeyPrefixReStakerAssetInfos reStakerId = clientChainAddr+'_'+ExoCoreChainIndex
-	// KeyPrefixReStakerAssetInfos key->value: reStakerId+'_'+AssetId->ReStakerSingleAssetInfo
-	// or reStakerId->mapping(AssetId->ReStakerSingleAssetInfo)?
+	// KeyPrefixReStakerAssetInfos reStakerID = clientChainAddr+'_'+ExoCoreChainIndex
```
restaker is a single word. should be restakerID.

rlan35 commented on x/delegation/keeper/keeper.go:
```diff
@@ -33,73 +32,49 @@ func NewKeeper(
-	return nil
+// GetExoCoreLzAppAddress Get exoCoreLzAppAddr from deposit keeper,it will be used when check the caller of precompile contract.
+// This function needs to be moved to `assets` module,which will facilitate its use for the other modules
+func (k *Keeper) GetExoCoreLzAppAddress(ctx sdk.Context) (common.Address, error) {
```
Exocore a single word

rlan35 commented on x/delegation/types/keys.go:
```diff
@@ -41,46 +38,42 @@ const (
 	// KeyPrefixDelegationUsedSalt key->value: operatorApproveAddr->map[salt]{}
 	KeyPrefixDelegationUsedSalt = []byte{prefixDelegationUsedSalt}
-	// KeyPrefixOperatorApprovedInfo key-value: operatorApproveAddr->map[reStakerId]{}
+	// KeyPrefixOperatorApprovedInfo key-value: operatorApproveAddr->map[reStakerID]{}
```
restakerID

TimmyExogenous committed:
fix the proto-lint error

TimmyExogenous commented:
All problems have been resolved. Additionally, the code related to setting `ExocoreLzAppAddress` has been refined.

MaxMustermann2 commented on precompiles/deposit/deposit_test.go:
```diff
@@ -52,12 +51,12 @@ func paddingClientChainAddress(input []byte, outputLength int) []byte {
-	exoCoreLzAppEventTopic := "0xc6a377bfc4eb120024a8ac08eef205be16b817020812c73223e81d1bdb9708ec"
-	usdtAddress := paddingClientChainAddress(common.FromHex("0xdAC17F958D2ee523a2206206994597C13D831ec7"), types.GeneralClientChainAddrLength)
-	usdcAddress := paddingClientChainAddress(common.FromHex("0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48"), types.GeneralClientChainAddrLength)
+	ExocoreLzAppAddress := "0x3fC91A3afd70395Cd496C647d5a6CC9D4B2b7FAD"
```
Was it your intention to capitalize the variable?

TimmyExogenous replied:
Thanks, It's a problem caused by being replaced globally. I will fix it.

MaxMustermann2 commented on precompiles/deposit/error.go:
```diff
@@ -2,6 +2,6 @@ package deposit
 
 const (
 	ErrContractInputParaOrType    = "the contract input parameter type or value error,arg index:%d, expected type is:%s,value:%v"
```
These errors are defined in multiple locations, why not consolidate them in one?

TimmyExogenous replied:
Yeah, we need a public directory in precompile to consolidate them.

MaxMustermann2 commented on proto/exocore/assets/v1/query.proto:
```diff
@@ -49,7 +50,7 @@ message QueryStakerAssetInfo {
 message QueryAssetInfoResponse {
   // asset_infos is the response for the staker asset info, indexed by the asset id.
-  map<string, StakerSingleAssetInfo> asset_infos = 1;
+  map<string, StakerAssetInfo> asset_infos = 1;
```
This is not a review comment but an insight into G705. Traditional maps should not be used in distributed systems because they are a source of non-determinism. However, protobuf internally sorts the maps by key when (de)-serializing them and hence such maps can be used. Note that the key must be of a simple sortable type for this to be done.

TimmyExogenous replied:
Yeah, you are right. It will cause a problem if the key isn't sortable.

MaxMustermann2 commented on proto/exocore/assets/v1/tx.proto:
```diff
@@ -137,9 +138,9 @@ message OperatorSingleAssetInfo {
 
-  // operator_own_amount is the amount that the operator owns.
+  // operator_amount is the amount that the operator owns.
   //todo: the field is used to mark operator's own assets and is not temporarily used now
```
To confirm, this todo still holds true right? An operator's self delegation is required, but it is treated the same way as any other delegation.

TimmyExogenous replied:
Yeah, the operator's self-delegation hasn't been completed yet. This field is added for forthcoming self-delegation implementation.

MaxMustermann2 commented on proto/exocore/assets/v1/tx.proto:
```diff
@@ -239,9 +240,29 @@ message RegisterAssetReq {
+  // authority is the address of the governance account.
+  string authority = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+
+  // params defines the x/evm parameters to update.
```
Not `x/evm` but `x/assets`

TimmyExogenous replied:
Thanks

MaxMustermann2 commented on proto/exocore/assets/v1/tx.proto:
```diff
@@ -239,9 +240,29 @@ message RegisterAssetReq {
 message RegisterAssetResponse {}
 
+// MsgUpdateParams is the Msg/UpdateParams request type for Erc20 parameters.
+// Since: cosmos-sdk 0.47
```
This `Since` can be removed since this module is not part of the Cosmos SDK.

TimmyExogenous replied:
Thanks

MaxMustermann2 commented on proto/exocore/assets/v1/tx.proto:
```diff
@@ -239,9 +240,29 @@ message RegisterAssetReq {
+
+// MsgUpdateParamsResponse defines the response structure for executing a
+// MsgUpdateParams message.
+// Since: cosmos-sdk 0.47
```
This `Since` can be removed since this module is not part of the Cosmos SDK.

MaxMustermann2 commented on x/operator/keeper/keeper.go:
```diff
@@ -16,10 +16,10 @@ type Keeper struct {
-	delegationKeeper     operatortypes.ExpectDelegationInterface
-	oracleKeeper         operatortypes.ExpectOracleInterface
-	avsKeeper            operatortypes.ExpectAvsInterface
+	assetsKeeper     keeper.Keeper
```
This needs to be an interface.

TimmyExogenous replied:
Thanks

TimmyExogenous replied:
This seems like it has been resolved.

MaxMustermann2 reviewed (commented):
There are still some unresolved conversations associated with prior reviews. Please resolve them or comment on them if you won't be fixing them.

TimmyExogenous committed:
refine the unconvitional code identified by code review in the procompile

TimmyExogenous committed:
add code comments for GetAvsSupportedAssets

TimmyExogenous committed:
Merge branch 'develop' into copy-app-chain

MaxMustermann2 commented on precompiles/delegation/error.go:
```diff
@@ -1,9 +1,5 @@
-	ErrCtxTxHash               = "ctx TxHash type error or is nil,type is:%v,value:%v"
-
-	ErrInputOperatorAddrLength = "mismatched length of the input operator address,actual is:%d,expect:%v"
+	ErrCtxTxHash = "ctx TxHash type error or is nil,type is:%v,value:%v"
```
Q: This is because the `UndelegationParams` object requires a transaction hash, which is not needed anywhere else. Is that correct?

TimmyExogenous replied:
Yeah, the record key of undelegation includes the TxHash. But it can't be accessed in the precompile, so I set it in the context when the EVM module processes the transaction.

MaxMustermann2 commented on x/operator/types/expected_keepers.go:
```diff
@@ -0,0 +1,112 @@
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type ExpectAssetsInterface interface {
```
Please rename to `AssetsKeeper`

TimmyExogenous replied:
Thanks

MaxMustermann2 commented on x/operator/keeper/dogfood.go:
```diff
@@ -0,0 +1,431 @@
+
+// IsOperatorOptingOutFromChainID returns true if the operator is opting out from the given
+// chain id.
+func (k *Keeper) IsOperatorOptingOutFromChainID(
```
Not a review comment, but in my PR I will remove this function and change the language around these in the next PR such that opt out completes immediately for the operator module.

TimmyExogenous replied:
Okay, thanks for your great work.

MaxMustermann2 commented on x/delegation/keeper/abci.go:
```diff
@@ -24,58 +20,80 @@ func (k Keeper) EndBlock(ctx sdk.Context, _ abci.RequestEndBlock) []abci.Validat
+			WaitUnbondingAmount: recordAmountNeg,
 		})
 		if err != nil {
 			panic(err)
```
Two points here:
- Please remove the `panic`s and use a cached context instead. If there is no error, the cached context's `writeFunc` should be called.
- Add a `TODO` here to say that the message should be sent to the client chain to release the assets for withdrawal. I spoke to @adu-web3 who said that the funds are made available for withdrawal after the undelegation is processed by Exocore. However, that is not the correct approach since the assets are locked for a 10-block (or larger) period during which they can be slashed.

TimmyExogenous replied:
> Please remove the panics and use a cached context instead. If there is no error, the cached context's writeFunc should be called.

The handle of errors in `BeginBlock` and `EndBlock` definitely needs to be reconsidered.  What's the meaning of using a cached context? This can be changed in the future PR.

> Add a TODO here to say that the message should be sent to the client chain to release the assets for withdrawal. I spoke to @adu-web3 who said that the funds are made available for withdrawal after the undelegation is processed by Exocore. However, that is not the correct approach since the assets are locked for a 10-block (or larger) period during which they can be slashed.

I remember the previously designed approach was that after undelegation is completed, the user's withdrawable status will increase. Exocore does not need to actively send notifications to the client chain about the completion of undelegation. The client chain will periodically or when the user initiates a transaction, synchronize the withdrawable status from Exocore to the client chain. Additionally, the code here is executed when the unbonding time expires, which means the related assets won't be slashed after this block.

MaxMustermann2 replied:
For the first point: understood, we will leave this for a future PR. A cached context allows for error handling within the SDK while not modifying the state.
```go
cachedCtx, writeCache := ctx.CacheContext()
if err := k.FunctionThatMightError(ctx); err != nil {
   return err
}
// dump events
ctx.EventManager().EmitEvents(cachedCtx.EventManager().Events())
// store the state change from the function
writeCache()
```

For the second point: The current implementation, in my understanding of the smart contract, is that whenever an undelegation request is executed on the client chain, a message reaches the Exocore chain. The message is executed in the `ExocoreGateway.sol`, which calls the undelegation precompile. The call's result (success or failure) is sent back to the client chain gateway. The client chain gateway then allows withdrawal of the undelegated assets, which is an incorrect implementation. The undelegated assets should remain stuck for 10 blocks (or the unbonding period, whichever is higher).

To fix this, my approach is to remove the immediate release feature from the client chain gateway, and for Exocore chain to send a message to the client chain when the assets are available. At this point, we may have to consider batching such messages. This is the TODO I was asking you to add here; however, it is not a blocker for this PR.

TimmyExogenous replied:
When the undelegation call's result is sent back to the client chain, It's just a message to notify the client chain that the undelegation has been successfully routed and queued in the Exocore chain. It doesn't mean the asset related to this undelegation can be withdrawn because the withdrawable amount increases only when this undelegation is completed.  As for how to notify the client chain after completing the undelegation, the client chain contract will sync the withdrawable state from the Exocore chain regularly or when a staker initiates actions(such as deposit, delegation, undelegation and withdrawing) from the client chain. If you find the withdrawable amount will increase immediately when
the undelegation call's result is successfully sent back to the client chain, there might be a bug in the client chain contract. Please let me know about it.

TimmyExogenous replied:
Appreciate for providing the cache context code.

MaxMustermann2 replied:
I misunderstood the Solidity implementation; please ignore the second part of my original comment.

MaxMustermann2 replied:
This comment is still left unresolved intentionally to remember the `panic` replacement.

TimmyExogenous committed:
resolve the missed issues identified by code review, such as adding some comments and refine the naming

TimmyExogenous committed:
fix the prtobuf lint error

MaxMustermann2 reviewed (approved):
Great job surviving through this long review process, I appreciate your patience!

There are a couple of questions / comments that I have left unresolved for future reference; they do not need to be answered or resolved now.

TimmyExogenous committed:
remove the unnecessary code related to operator from the delegation proto file

adu-web3 commented:
> The message is executed in the ExocoreGateway.sol, which calls the undelegation precompile. The call's result (success or failure) is sent back to the client chain gateway. The client chain gateway then allows withdrawal of the undelegated assets, which is an incorrect implementation

@MaxMustermann2 this is not the case for current implementation, every withdrawal must be processed by Exocore network so client chain contracts can not unlock the asset and allow for further claim by itself. For undelegation workflow, the call's result of ExocoreGateway could only indicate whether the undelegation is queued to be finished or not. We could change the result's name in contract to eliminate misunderstanding

leonz789 commented:
`ginkgo` used across all _test file, but it seems only used for evmos/inflation module's integration test, all files imported ginkgo here actually not use it(no any `Describe` or other ginkgo test usage). Is there any consideration to use ginkgo/gomega ?

MaxMustermann2 commented:
> @MaxMustermann2 this is not the case for current implementation, every withdrawal must be processed by Exocore network so client chain contracts can not unlock the asset and allow for further claim by itself. For undelegation workflow, the call's result of ExocoreGateway could only indicate whether the undelegation is queued to be finished or not. We could change the result's name in contract to eliminate misunderstanding

I see; this was a mistake in my understanding of the Solidity implementation. 

- User undelegates and a message is sent to Exocore chain. A response is returned indicating the success or failure of the undelegation request on the Exocore chain. This response results in an event being emitted on the client chain and nothing else.
- When the undelegation matures, the undelegation is marked on Exocore chain as ready for withdrawal.
- The user then sends a withdraw transaction on the client chain, which is forwarded to the Exocore chain. The Exocore chain returns the amount that can be withdrawn.
- This amount is updated in the `Vault` contract, and then user needs to withdraw from the Vault.

Thank you for the clarification.

TimmyExogenous commented:
> `ginkgo` used across all _test file, but it seems only used for evmos/inflation module's integration test, all files imported ginkgo here actually not use it(no any `Describe` or other ginkgo test usage). Is there any consideration to use ginkgo/gomega ?

The functions `RegisterFailHandler` and `RunSpecs` will use them.

leonz789 commented:
> > `ginkgo` used across all _test file, but it seems only used for evmos/inflation module's integration test, all files imported ginkgo here actually not use it(no any `Describe` or other ginkgo test usage). Is there any consideration to use ginkgo/gomega ?
> 
> The functions `RegisterFailHandler` and `RunSpecs` will use them.

yes, but those two also meaningless, there's nothing for `RunSpecs` to run

mikebraver commented on proto/exocore/assets/v1/tx.proto:
```diff
@@ -203,9 +240,27 @@ message RegisterAssetReq {
+
+// MsgUpdateParamsResponse defines the response structure for executing a
+// MsgUpdateParams message.
+message MsgUpdateParamsResponse {}
```
Why didn't add fields in response message?

TimmyExogenous replied:
No data needs to be returned, so an empty struct is defined as the response, with reference to cosmos-sdk.

TimmyExogenous commented:
> > > `ginkgo` used across all _test file, but it seems only used for evmos/inflation module's integration test, all files imported ginkgo here actually not use it(no any `Describe` or other ginkgo test usage). Is there any consideration to use ginkgo/gomega ?
> > 
> > 
> > The functions `RegisterFailHandler` and `RunSpecs` will use them.
> 
> yes, but those two also meaningless, there's nothing for `RunSpecs` to run

I haven't thoroughly researched it. It might be used to collect the results when running the tests. The output log will be different when I run the tests with it. It's fine to remove it if it doesn't influence the test process.

leonz789 commented:
> > > > `ginkgo` used across all _test file, but it seems only used for evmos/inflation module's integration test, all files imported ginkgo here actually not use it(no any `Describe` or other ginkgo test usage). Is there any consideration to use ginkgo/gomega ?
> > > 
> > > 
> > > The functions `RegisterFailHandler` and `RunSpecs` will use them.
> > 
> > 
> > yes, but those two also meaningless, there's nothing for `RunSpecs` to run
> 
> I haven't thoroughly researched it. It might be used to collect the results when running the tests. The output log will be different when I run the tests with it. It's fine to remove it if it doesn't influence the test process.

for the output log print by `ginkgo`, you can see all the statistics is 0 (passed, failed, skipped), since there's no actual test case related to it
